GVRf/Framework/framework/src/main/java/org/gearvrf/io/GearCursorController.java

```diff
@@ -533,16 +571,24 @@ void handleControllerEvent(final ControllerEvent event) {
                                         prevButtonHome, KeyEvent.KEYCODE_HOME);
             prevButtonHome = handleResult == -1 ? prevButtonHome : handleResult;
             event.recycle();
-            if (mSendEventsToActivity)
+            if (mSendEventsToActivity && ((keyEvent.size() > 0) || (motionEvent.size() > 0)))
             {
-                GVRActivity activity = getGVRContext().getActivity();
-                for (KeyEvent e : keyEvent)
+                mPropagateEvents.init(keyEvent, motionEvent);
+                if (Looper.getMainLooper().equals(Looper.myLooper()))
                 {
-                    activity.dispatchKeyEvent(e);
+                    mPropagateEvents.run();
```

Comment:
What is mPropagateEvents for? It seems to me like both branches of this if effectively do the same thing on the same thread.

Reply:
It crashes in some applications when I call GVRActivity.dispatchMotionEvent. I will go back and figure out which one crashed. Maybe there is a better way to handle it. I was getting "you are doing stuff on the wrong thread" messages from Android.

Reply:
I checked into it and this code can go. The crash happens in gvr-events. GVRActivity.dispatchKeyEvent is called from the controller thread and it causes this exception: 
android.view.ViewRootImpl$CalledFromWrongThreadException: Only the original thread that created a view hierarchy can touch its views.

Reply:
See Activity.runOnUiThread: https://developer.android.com/reference/android/app/Activity.html#runOnUiThread(java.lang.Runnable). Probably the sample itself needs to place the call on the right thread; not GVRf's responsibility what happens in an event handler.

Reply:
I took a look at the logic and you are right - it sucks. I recoded it to always propagate the events to the UI thread and now gvr-events works.

Reply:
The Gear controller thread will NEVER be the UI thread and usually you want to respond to these events on that thread. I had the comparison backwards.